rules/aws_s3_bucket_name.go

```diff
@@ -51,6 +64,68 @@ func (r *AwsS3BucketNameRule) Link() string {
 	return project.ReferenceLink(r.Name())
 }
 
+// Return regex bucket naming rules
+func rules() ([]map[string]string, error) {
+	// Get bucket naming rule based on length from AWS SDK API models
+	r, err := http.Get(
+		"https://raw.githubusercontent.com/aws/aws-sdk-go/main/models/apis/cloudformation/2010-05-15/api-2.json",
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Body.Close()
+
+	bucket := Bucket{}
+	err = json.NewDecoder(r.Body).Decode(&bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	// https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
+	return []map[string]string{
```

Comment:
Use `struct` when fields are known and reserve `map` for arbitrary keys. Avoid unnecessary shorthand and rely on the fact that defined keys allows quick auto-completion.

```go
type struct S3BucketNameRule {
  Regexp regexp.Regexp
  Description string
}
```

If you take out the network component you don't need to return an error here so it's unclear that `rules()` is even needed as its own function. So you could potentially just define this as inline anonymous struct type:

```go
rules := []struct{
  Regexp regexp.Regexp
  Description string
}{
  {
  // ...
  }
}
```

Reply:
Additionally, if you take out the network component it should be simpler to define these as a global variable and use `regexp.MustCompile`. It's better to fail at the initialization level and effectively break the build than risk an unexpected runtime error.

Reply:
I'd remove the function and just add this  inline inside _Check_:
```go
bucketNameMinLength := 3
bucketNameMaxLength := 63

// https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
rules := []struct {
  Regexp      regexp.Regexp
  Description string
}{
  {
    Regexp: *regexp.MustCompile(fmt.Sprintf(
      "^(.{0,%d})?(.{%d,})?$",
      bucketNameMinLength-1,
      bucketNameMaxLength+1,
    )),
    Description: fmt.Sprintf(
      "Bucket names must be between %d (min) and %d (max) characters long.",
      bucketNameMinLength,
      bucketNameMaxLength,
    ),
  },
  {
    Regexp:      *regexp.MustCompile("[^a-z0-9\\.\\-]"),
    Description: "Bucket names can consist only of lowercase letters, numbers, dots (.), and hyphens (-).",
  },

  ...

}
```
Or is it preferable to write this another place?

And then:
```go
regex, _ := regexp.MatchString(rule.Regexp.String(), name)
if regex {
  runner.EmitIssue(
    r,
    fmt.Sprintf(`%s (name: %q, regex: %q)`, rule.Description, name, rule.Regexp.String()),
    attribute.Expr.Range(),
  )
}
```

Reply:
Inline is fine. Avoid `regexp.MatchString` in favor of compiling the regex first and then testing the string separately.

Reply:
This isn't needed:

```go
regex, err = regexp.Compile(rule.Regexp.String())
```

`rule.Regexp` is itself already a compiled `regexp.Regexp` object. I only mean to avoid the `regexp.MatchString` function which accepts a string regex and string value and both compiles the regex and matches the string. Calling the `MatchString` _method_ on an already compiled regex is fine.

I'm happy to tweak this a bit to fine tune where in the program lifecycle the regexes are compiled.

Reply:
So the for loop already sets the right compiled regex from the current iterator item?
```go
for _, rule := range regexRules {
  if rule.Regexp.MatchString(name) {
    runner.EmitIssue(
      r,
      fmt.Sprintf(`%s (name: %q, regex: %q)`, rule.Description, name, rule.Regexp.String()),
      attribute.Expr.Range(),
    )
  }
}
```

Reply:
Correct!